crates/transport/src/actors/portal_logs_collector.rs

```diff
@@ -0,0 +1,181 @@
+use std::{sync::Arc, time::Duration};
+
+use futures::StreamExt;
+use futures_core::Stream;
+use libp2p::{
+    kad::RecordKey, request_response::ResponseChannel, swarm::{NetworkBehaviour, SwarmEvent, ToSwarm}, PeerId, Swarm
+};
+use libp2p_swarm_derive::NetworkBehaviour;
+use serde::{Deserialize, Serialize};
+use tokio_util::sync::CancellationToken;
+
+use sqd_messages::QueryFinished;
+
+use crate::{
+    behaviour::{
+        base::{BaseBehaviour, BaseBehaviourEvent},
+        request_server::{Request, ServerBehaviour},
+        wrapped::{BehaviourWrapper, TToSwarm, Wrapped},
+    },
+    codec::ProtoCodec,
+    protocol::{
+        MAX_QUERY_MSG_SIZE, MAX_QUERY_RESULT_SIZE, PORTAL_LOGS_PROTOCOL, PORTAL_LOGS_PROVIDER_KEY,
+    },
+    record_event,
+    util::{new_queue, Sender, TaskManager, DEFAULT_SHUTDOWN_TIMEOUT},
+};
+
+#[derive(Debug)]
+pub enum PortalLogsCollectorEvent {
+    LogQuery { peer_id: PeerId, log: QueryFinished },
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct PortalLogsCollectorConfig {
+    pub events_queue_size: usize,
+    pub shutdown_timeout: Duration,
+}
+
+impl PortalLogsCollectorConfig {
+    pub fn new() -> Self {
+        Self {
+            events_queue_size: 100,
+            shutdown_timeout: DEFAULT_SHUTDOWN_TIMEOUT,
+        }
+    }
+}
+
+type PortalLogCollectionBehaviour = Wrapped<ServerBehaviour<ProtoCodec<QueryFinished, ()>>>;
+
+#[derive(NetworkBehaviour)]
+pub struct InnerBehaviour {
+    base: Wrapped<BaseBehaviour>,
+    collector: PortalLogCollectionBehaviour,
+}
+
+pub struct PortalLogsCollectorBehaviour {
+    inner: InnerBehaviour,
+}
+
+impl PortalLogsCollectorBehaviour {
+    pub fn new(mut base: BaseBehaviour) -> Wrapped<Self> {
+        base.set_server_mode();
+        let _ = base
+            .get_kademlia_mut()
+            .start_providing(RecordKey::new(PORTAL_LOGS_PROVIDER_KEY));
+        log::error!("Start providing: {:?}", PORTAL_LOGS_PROVIDER_KEY);
+        Self {
+            inner: InnerBehaviour {
+                base: base.into(),
+                collector: ServerBehaviour::new(
+                    ProtoCodec::new(MAX_QUERY_MSG_SIZE, MAX_QUERY_RESULT_SIZE),
+                    PORTAL_LOGS_PROTOCOL,
+                    Duration::from_secs(5),
```

Comment:
Even if you want to make it hard-coded, let's put it in a constant like `REQUEST_TIMEOUT`